File: packages/core/src/projects/scope.ts
import type { ProjectIdentityRef, SessionHead } from "../types/index.js";
import {
  computeIdentity,
  matchesProjectIdentity,
  normalizeProjectDirectory,
  type IdentityFs,
} from "./identity.js";
import { realFs } from "./fs.js";

export interface ProjectScopeMatcher {
  identity: ProjectIdentityRef;
  path: string;
}

export function createProjectScopeMatcher(
  queryPath: string,
  fs: IdentityFs = realFs,
): ProjectScopeMatcher {
  const identity = computeIdentity(queryPath, fs);
  return createProjectScopeMatcherFromIdentity(queryPath, identity);
}

export function createProjectScopeMatcherFromIdentity(
  queryPath: string,
  identity: ProjectIdentityRef,
): ProjectScopeMatcher {
  return {
    identity: { kind: identity.kind, key: identity.key },
    path: normalizeProjectScopePath(queryPath),
  };
}

export function matchesProjectScope(session: SessionHead, scope: ProjectScopeMatcher): boolean {
  if (matchesProjectIdentity(session.project_identity, scope.identity)) return true;
  return session.directory ? isPathScopeMatch(scope.path, session.directory) : false;
}

export function filterSessionsByProjectScope<T extends SessionHead>(
  sessions: T[],
  queryPath: string,
  fs?: IdentityFs,
): T[] {
  const scope = createProjectScopeMatcher(queryPath, fs);
  return sessions.filter((session) => matchesProjectScope(session, scope));
}

function isPathScopeMatch(queryPath: string, sessionPath: string): boolean {
  const session = normalizeProjectScopePath(sessionPath);
  return (
    session === queryPath ||
    session.startsWith(queryPath + "/") ||
    queryPath.startsWith(session + "/")
  );
}

export function normalizeProjectScopePath(path: string): string {
  return normalizeProjectDirectory(path).replaceAll("\\", "/");
}

File: crates/codesesh-core/src/projects/scope.rs
use super::{compute_identity, matches_project_identity, normalize_project_directory};
use crate::contract::{ProjectIdentity, SessionHead};

#[derive(Clone, Debug)]
pub struct ProjectScopeMatcher {
    pub identity: ProjectIdentity,
    pub path: String,
}
pub fn normalize_project_scope_path(path: &str) -> String {
    normalize_project_directory(path).replace('\\', "/")
}
pub fn create_project_scope_matcher(path: &str) -> ProjectScopeMatcher {
    create_project_scope_matcher_from_identity(path, compute_identity(path))
}
pub fn create_project_scope_matcher_from_identity(
    path: &str,
    identity: ProjectIdentity,
) -> ProjectScopeMatcher {
    ProjectScopeMatcher {
        identity,
        path: normalize_project_scope_path(path),
    }
}
pub fn matches_project_scope(session: &SessionHead, scope: &ProjectScopeMatcher) -> bool {
    if matches_project_identity(&session.project_identity, &scope.identity) {
        return true;
    }
    if session.directory.is_empty() {
        return false;
    }
    let path = normalize_project_scope_path(&session.directory);
    path == scope.path
        || path.starts_with(&format!("{}/", scope.path))
        || scope.path.starts_with(&format!("{path}/"))
}
pub fn filter_sessions_by_project_scope<'a>(
    sessions: &'a [SessionHead],
    query_path: &str,
) -> Vec<&'a SessionHead> {
    let scope = create_project_scope_matcher(query_path);
    sessions
        .iter()
        .filter(|session| matches_project_scope(session, &scope))
        .collect()
}
